### DISCUSSION TITLE
Confusion about the result of bool type computing

### DISCUSSION THREAD

**Discussion by Vernlium** [halide/Halide#6813] [Q&A]
Hi all.

In Halide 14.0.0, the behavior of calculation about “bool” is very weird.

We have `Func0(x,y)` as type bool and tried:

```c++
Func1(x,y) = Func0(x,y) + 4;
Func1(x,y) = Func0(x,y) * 4;
Func1(x,y) = 4/Func0(x,y);
Func1(x,y) = Func0(x,y) << 4;
```

We get error return is:

```
Unhandled exception: Error: Integer constant 4 will be implicitly coerced to type uint1, which changes its value to (uint1)0
```

But when we tried:

```c++
Func1(x,y) = Func0(x,y) + 4.f;
Func1(x,y) = Func0(x,y) + Funcs();  //Here we define Funcs() as integer
Func1(x,y) = pow(4, Func0(x,y));
Func1(x,y) = pow(Func0(x,y), 4);
```

They all works well.

So, we would like to know why was that? Shouldn’t their behavior have some kind of consistency?

And the output type of “bool” calculation is also very weird, could anyone give the regular pattern?

We have equation as below:

```c++
Func3(x,y) = pow(Func2(x, y), Func1(x, y));
```

Here, 
If Func2 is float16 and Func1 is bool, Func3 will be float16.
If Func2 is bool and Func1 is float16, Func3 will be float32.
If Func2 is bool and Func1 is int16, Func3 will be float32.
If Func2 is int16 and Func1 is bool, Func3 will be float32.

And the most weird is that if Func2 is bool and Func1 is bool, Func3 will be float32.

Simply saying , we get:

```
Func2(fp16), Func1(bool) –> Func3(fp16)
Func2(bool), Func1(fp16) –> Func3(fp32)
Func2(bool), Func1(int16) –> Func3(fp32)
Func2(int16), Func1(bool) –> Func3(fp32)
Func2(bool), Func1(bool) –> Func3(fp32)
```

And we consider one input is scalar,  if the expr is

```
Func3(x,y) = pow(4, Func1(x, y))
```

and set Func1 to bool, we get Func3 is float32. 

But if the expr is

```
Func3(x,y) = pow(Func1(x, y), 4)
```

and Func1 is bool , we get Func3 is bool.

 And also if the expr is

```
Func3(x,y) = pow(Func1(x, y), Func2())
```

setting Func1 is bool and Func2 is int16, we get Func3 is fp32.

Simply, we get:

```
4, Func1(bool) -> Func3(fp32)
Func1(bool), 4 -> Func3(bool)
Func1(bool), Func2(scalar with type int16) -> Func3(fp32)
```

The inference of bool type calculation is so untellable which make us have to test before using. What rule about them? Or is it possible to make them more straightforward?

Looking forward for your reply, best wishes!

**Comment by steven-johnson** [comment]
This sounds like a bug to me. Opened https://github.com/halide/Halide/issues/6820 for tracking. Thanks for pointing this out!

**Comment by abadams** [comment]
The short answer is: 
- Bool is treated as a uint1 type. 
- pow matches the type of the first operand if the second operation is an int literal. Otherwise it returns a float of some appropriate width.
